client/client.go

```diff
@@ -353,94 +354,75 @@ func (kc *kClient) applyCallOptions(ctx context.Context, cfg rpcinfo.MutableRPCC
 }
 
 // Call implements the Client interface .
-func (kc *kClient) Call(ctx context.Context, method string, request, response interface{}) error {
-	if !kc.inited {
-		panic("client not initialized")
-	}
-	if kc.closed {
-		panic("client is already closed")
-	}
-	if ctx == nil {
-		panic("ctx is nil")
-	}
+func (kc *kClient) Call(ctx context.Context, method string, request, response interface{}) (err error) {
+	kc.validateForCall(ctx)
 	var ri rpcinfo.RPCInfo
 	var callOpts *callopt.CallOptions
 	ctx, ri, callOpts = kc.initRPCInfo(ctx, method)
 
 	ctx = kc.opt.TracerCtl.DoStart(ctx, ri)
-
-	var callOptRetry retry.Policy
-	var callOptFallback *fallback.Policy
-	if callOpts != nil {
-		callOptFallback = callOpts.Fallback
-		if callOpts.RetryPolicy.Enable {
-			callOptRetry = callOpts.RetryPolicy
+	var reportErr error
+	var recycleRI bool
+	defer func() {
+		if panicInfo := recover(); panicInfo != nil {
+			err = stats.ClientPanicToErr(ctx, panicInfo, ri, false)
+			reportErr = err
 		}
-	}
+		kc.opt.TracerCtl.DoFinish(ctx, ri, reportErr)
+		if recycleRI {
+			// why need check recycleRI to decide if recycle RPCInfo?
+			// 1. no retry, rpc timeout happen will cause panic when response return
+			// 2. retry success, will cause panic when first call return
+			// 3. backup request may cause panic, cannot recycle first RPCInfo
+			// RPCInfo will be recycled after rpc is finished,
+			// holding RPCInfo in a new goroutine is forbidden.
+			rpcinfo.PutRPCInfo(ri)
+		}
+		callOpts.Recycle()
+	}()
+
+	callOptRetry, callOptFallback := initRetryAndFallbackWithCallopts(callOpts)
```

Comment:
callOptRetry is not a pointer to `retry.Policy`. It's a rather large struct and may be copied 5 times if enabled.

Reply:
btw, I prefer to separate asignment of these two callOpts, since callOptFallback is only used in doFallbackIfNeeded, we can get the value in `getFallbackPolicy`

Reply:
good suggestion, done